### PR TITLE
Switching KubeVirt platform's default ingress to NodePort

### PR DESCRIPTION
**Pull request by davidvossel**
Previously, Clusters using the KubeVirt platform would get stuck at a partial install. This was primarily due ingress routing failures which prevented console from working and also prevented guest cluster API access.

Our ingress issues are mostly resolved by switching from HostNetwork to NodePort for the guest cluster
s default ingress. Before this change, when we used HostNetwork for ingress, we didn't have a way to accurately select which VMs were running the HA proxy instances when routing default ingress traffic. This made it difficult to target the correct VMs when routing traffic. NodePort allows us to target all VMs when routing traffic.

By switching to NodePort for ingress and exposing the cluster API using routes and a LB ( #1032) we now have the ability to completely install an ocp tenant cluster using the KubeVirt platform. The e2e tests have now been updated to fully exercise the guest api access and verify a complete install.

## NOTE e2e test wildcard *.apps routing

In order to avoid needing to create a LB+DNS entry for the KubeVirt cluster's default ingress, we have deployed a trick in the e2e tests which leverages the mgmt cluster's existing LB and wildcard *.apps DNS entry.

It works like this.

Let's say the mgmt cluster's default ingress is DNS record is`*.apps.mgmt-cluster.example.com`. If we use the base domain `apps.mgmt-cluster.example.com` for our tenant cluster, then the tenant cluster's default ingress will become something like `*.apps.tenant-cluster.apps.mgmt-cluster.example.com`. By making the tenant cluster's default ingress a subdomain of the mgmt cluster's wildcard DNS record, we can use the mgmt cluster's existing ingress to route to the tenant cluster's default ingress.

In this example, our e2e tests get ingress for the tenant cluster to work by simply creating a subdomain route on the mgmt cluster that routes everything from `*.apps.tenant-cluster.apps.mgmt-cluster.example.com` on the mgmt cluster to the kubevirt VMs. see the ```CreateKubeVirtClusterWildcardRoute``` function in this PR for more details.

In production, someone could deploy a similar trick to this if they wanted, or they could create an external LB + DNS entry which routes to all the kubevirt VMs in the tenant cluster. Either way works now that we're using NodePort and all VMs are routeable for ingress.

As a follow up to this PR, the quickstart guide will be updated to document the different methods for getting default ingress to work for KubeVirt tenant clusters.

### CODE REVIEW

File: control-plane-operator/hostedclusterconfigoperator/controllers/resources/registry/registry.go [hunk 1]
```diff
@@ -26,7 +27,7 @@ func ReconcileRegistryConfig(cfg *imageregistryv1.Config, isPlatformNone bool) {
 	cfg.Spec.Requests.Write.MaxRunning = 0
 	cfg.Spec.Requests.Write.MaxWaitInQueue.Reset()
 
-	if isPlatformNone {
+	if platform == hyperv1.KubevirtPlatform || platform == hyperv1.NonePlatform {
```
Comment (enxebre) [inline control-plane-operator/hostedclusterconfigoperator/controllers/resources/registry/registry.go:30]: Why this change?
How is the change related to this PR? "Switching KubeVirt platform's default ingress to NodePort"
We should possibly run once and stop ReconcileRegistryConfig so user has the ability configure the registry not only for IBM, so unless there's a reason for this change to be here, I'd suggest to have a separate PR discussion.

File: support/globalconfig/infrastructure.go [hunk 2]
```diff
@@ -21,18 +21,26 @@ func InfrastructureConfig() *configv1.Infrastructure {
 
 func ReconcileInfrastructure(infra *configv1.Infrastructure, hcp *hyperv1.HostedControlPlane) {
 
+	platformType := hcp.Spec.Platform.Type
+	switch hcp.Spec.Platform.Type {
+	// In case of KubeVirtPlatform, set the Infrastructure to NonePlatform
```
Comment (enxebre) [inline support/globalconfig/infrastructure.go:26]: can you elaborate here? which mcs error and why?

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 3]
```diff
@@ -1145,7 +1145,7 @@ func reconcileHostedControlPlane(hcp *hyperv1.HostedControlPlane, hcluster *hype
 	// Pass through Platform spec.
 	hcp.Spec.Platform = *hcluster.Spec.Platform.DeepCopy()
 	switch hcluster.Spec.Platform.Type {
-	case hyperv1.AgentPlatform, hyperv1.KubevirtPlatform:
+	case hyperv1.AgentPlatform:
```
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:1148]: can you clarify how does this make a difference?

**Comment by enxebre** [comment]
can we now update the e2e and address all the kubevirt TODOs? e.g // TODO: Replace this with the generic WaitForNReadyNodes() function?

**Comment by davidvossel** [comment]
>can we now update the e2e and address all the kubevirt TODOs? e.g // TODO: Replace this with the generic WaitForNReadyNodes() function and align with AWS coverage?

yes, see the changes in `test/e2e/create_cluster_test.go`. The TODO comments have been replaced by the usage of WaitForNReadyNodes and WaitForImageRollout.

File: support/globalconfig/infrastructure.go (hunk 2 above)
Comment (davidvossel) [reply support/globalconfig/infrastructure.go:26]: A few years ago there was an effort to make an ipi installer for KubeVirt. This was abandoned, however some logic related to the kubevirt platform remains in the mco/mcs. Unfortunately the left over IPI logic related to kubevirt causes the mcs not to properly serve up ignition when the mcs/mco detects the KubeVirt platform on the `infrastructure` object instead of the previous `none` object.

We will sort through this on the mcs side eventually, for now this issue blocks us from being able to transition to NodePort unless we do an override here for the infrastructure object and keep using `none` when kubevirt is in use.

**Comment by davidvossel** [comment]
the kubevirt CI lane will fail until this is merged, https://github.com/openshift/release/pull/27060

File: support/globalconfig/infrastructure.go (hunk 2 above)
Comment (enxebre) [reply support/globalconfig/infrastructure.go:26]: Thanks! make sense, could you please elaborate that a bit in the `// comment`? e.g "the mcs won't serve ignition when configv1.InfrastructureStatus.PlatformStatus is kubevirt so we set it to None" and point to the relevant issue tracking this?

**Comment by enxebre** [comment]
Thanks @davidvossel!

>There is some indirection going on here. I'll explain... it's frustrating.

How can we help to reduce the hassle? Do you think this is reasonable result from current organic grow and being figuring the best setup for each platform or do you have other concerns?

/lgtm

**Comment by davidvossel** [comment]
/test e2e-kubevirt-gcp

hopefully this works now that the prow config uses the latest operator in this PR for the CPO and HCCO

**Comment by davidvossel** [comment]
yay kubevirt lane passed. We now have ingress working!

**Comment by davidvossel** [comment]
>How can we help to reduce the hassle? Do you think this is reasonable result from current organic grow and being figuring the best setup for each platform or do you have other concerns?

@enxebre I think this is simply the nature of the project and what we're all building.

The difficulty and indirection uncovered in this PR was ultimately our (folks working on kubevirt platform) responsibility. We took a short cut by leveraging the `none` platfrom's CPO/HCCO logic. That helped us make incremental progress early on, but meant we had to unpack more complexity later on when we needed to touch the CPO/HCCO for custom kubevirt logic.

so, we did this to ourselves, but I still think it was a good trade off as it allowed us to get bootstrapped quickly.
